[Wasm-GC] Add support for new constant expressions

takikawa opened the pull request:
#### 2b9318e65c51add3e3d798360a05960524b9fa92
<pre>
[Wasm-GC] Add support for new constant expressions
<a href="https://bugs.webkit.org/show_bug.cgi?id=254696">https://bugs.webkit.org/show_bug.cgi?id=254696</a>

Reviewed by Justin Michaud.

Enables new constant expressions used in the GC proposal: array.new (normal,
default, and fixed), struct.new (normal, default),
extern.internalize/externalize, and i31.new. Both GC and extended constant expr
flags need to be turned on to use these.

The representation of ExpressionType used in the constant expression generator
is changed to potentially hold a Strong&lt;JSObject&gt; reference in order to keep
allocated objects live.

This enables most GC array spec tests to run, so this patch also adds
those tests to the suite. The remaining test failure is blocked on
parsing element vector items.

Running these spec tests also required updating part of the GC spec test
harness to match the upstream JS test helpers.

This patch also fixes a minor bug in parsing `array.new_data` that&apos;s exposed by
these tests.

* JSTests/wasm/gc-spec-harness.js:
* JSTests/wasm/gc-spec-harness/async_index.js:
(hostref):
(eq_ref):
(reinitializeRegistry.get promise_test):
(reinitializeRegistry):
(externref): Deleted.
(is_externref): Deleted.
(is_funcref): Deleted.
(eq_externref): Deleted.
(eq_funcref): Deleted.
* JSTests/wasm/gc-spec-harness/sync_index.js:
(hostref):
(eq_ref):
(externref): Deleted.
(is_externref): Deleted.
(is_funcref): Deleted.
(eq_externref): Deleted.
(eq_funcref): Deleted.
* JSTests/wasm/gc-spec-harness/wasm-constants.js: Added.
(bytes):
(bytesWithHeader):
(makeSig):
(makeSig_v_x):
(makeSig_v_xx):
(makeSig_r_v):
(makeSig_r_x):
(makeSig_r_xx):
(assertTraps):
(assertWasmThrows):
* JSTests/wasm/gc-spec-tests/array.wast.js: Added.
* JSTests/wasm/gc-spec-tests/i31.wast.js: Added.
* JSTests/wasm/gc/const-exprs-flag-off.js: Added.
(async testInvalidGCConstExprs):
* JSTests/wasm/gc/const-exprs.js: Added.
(async testGCConstExprs):
(async testInvalidConstExprs):
* Source/JavaScriptCore/wasm/WasmConstExprGenerator.cpp:
(JSC::Wasm::ConstExprGenerator::ConstExprValue::ConstExprValue):
(JSC::Wasm::ConstExprGenerator::ConstExprValue::getValue):
(JSC::Wasm::ConstExprGenerator::ConstExprValue::type):
(JSC::Wasm::ConstExprGenerator::ConstExprValue::operator+):
(JSC::Wasm::ConstExprGenerator::ConstExprValue::operator-):
(JSC::Wasm::ConstExprGenerator::ConstExprValue::operator*):
(JSC::Wasm::ConstExprGenerator::result const):
(JSC::Wasm::ConstExprGenerator::addConstant):
(JSC::Wasm::ConstExprGenerator::getGlobal):
(JSC::Wasm::ConstExprGenerator::addRefI31):
(JSC::Wasm::ConstExprGenerator::createNewArray):
(JSC::Wasm::ConstExprGenerator::addArrayNew):
(JSC::Wasm::ConstExprGenerator::addArrayNewDefault):
(JSC::Wasm::ConstExprGenerator::addArrayNewFixed):
(JSC::Wasm::ConstExprGenerator::createNewStruct):
(JSC::Wasm::ConstExprGenerator::addStructNewDefault):
(JSC::Wasm::ConstExprGenerator::addStructNew):
(JSC::Wasm::ConstExprGenerator::addAnyConvertExtern):
(JSC::Wasm::ConstExprGenerator::addExternConvertAny):
(JSC::Wasm::ConstExprGenerator::addRefFunc):
(JSC::Wasm::evaluateExtendedConstExpr):
* Source/JavaScriptCore/wasm/WasmConstExprGenerator.h:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseInitExpr):
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: <a href="https://commits.webkit.org/269484@main">https://commits.webkit.org/269484@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5dabf3555ef8b28fe25a78fb8f629b0b3644251f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22678 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/358 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23765 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24587 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20997 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22937 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1431 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23208 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21936 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22918 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19681 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25440 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20552 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26786 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19752 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20795 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24623 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22044 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/229 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18083 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/26096 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6118 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5404 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/27378 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/238 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5944 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmConstExprGenerator.cpp:
```diff
@@ -188,25 +192,121 @@ class ConstExprGenerator {
+        WASM_PARSER_FAIL_IF(!Options::useWebAssemblyGC(), "Wasm GC is not enabled");
+
+        if (m_mode == Mode::Evaluate)
+            result = createNewArray(typeIndex, static_cast<uint32_t>(size), value);
```
Are we actually guaranteed that these will stay alive? When they are on the stack or in a register, they are fine, but what if they get stored only to a ResultList (which is a Vector<>)?

takikawa replied:
Hi @justinmichaud is the situation that you are imagining something like the following?

```
(module
  (type (struct))
  (global (ref 0) (struct.new 0) (struct.new 0) (struct.new 0)) ;; doesn't validate, but suppose it does
)
```

In the global, there are three struct instances on the stack in the `ResultList`. This example doesn't validate, but let's say it did and only the last item on the stack was taken as the result. Then the other two structs are not reachable and get deallocated after the const expr generator is destroyed.

I think in practice, only init expressions where allocated values will be kept live pass validation. In particular, init expressions don't have `drop` so any allocated values must either be consumed (to be put in another object's fields) or returned from the init expr (and thus have a reference to them). And init exprs can't do any mutation of state, so you can't launder a reference out either.

Might be misunderstanding the concern though, happy to look into this more.

justinmichaud reviewed (approved):
r=me, but I am pretty sure this is going to have UAF problems

takikawa commented on Source/JavaScriptCore/wasm/WasmConstExprGenerator.cpp:
```diff
@@ -43,10 +47,70 @@ class ConstExprGenerator {
 
+    // Represents values that a constant expression may evaluate to.
+    // If a constant expression allocates an object, it should be put in a Strong handle.
+    struct ConstExprValue {
```
I believed I've fixed the GC issues now, by changing `ExpressionType` to use a new struct `ConstExprValue`. It can either hold bits for a non-object or a `Strong<JSObject>` handle for keeping an allocated object alive.

I wanted to write a specific independent test that triggered the GC failure, but it seems the best way to trigger the failure is to run the existing tests in the continuous-gc mode (these should now pass too).

justinmichaud reviewed (approved):
r=me, sweet!

takikawa committed:
[Wasm-GC] Add support for new constant expressions
https://bugs.webkit.org/show_bug.cgi?id=254696

Reviewed by Justin Michaud.

Enables new constant expressions used in the GC proposal: array.new (normal,
default, and fixed), struct.new (normal, default),
extern.internalize/externalize, and i31.new. Both GC and extended constant expr
flags need to be turned on to use these.

The representation of ExpressionType used in the constant expression generator
is changed to potentially hold a Strong<JSObject> reference in order to keep
allocated objects live.

This enables most GC array spec tests to run, so this patch also adds
those tests to the suite. The remaining test failure is blocked on
parsing element vector items.

Running these spec tests also required updating part of the GC spec test
harness to match the upstream JS test helpers.

This patch also fixes a minor bug in parsing `array.new_data` that's exposed by
these tests.

* JSTests/wasm/gc-spec-harness.js:
* JSTests/wasm/gc-spec-harness/async_index.js:
(hostref):
(eq_ref):
(reinitializeRegistry.get promise_test):
(reinitializeRegistry):
(externref): Deleted.
(is_externref): Deleted.
(is_funcref): Deleted.
(eq_externref): Deleted.
(eq_funcref): Deleted.
* JSTests/wasm/gc-spec-harness/sync_index.js:
(hostref):
(eq_ref):
(externref): Deleted.
(is_externref): Deleted.
(is_funcref): Deleted.
(eq_externref): Deleted.
(eq_funcref): Deleted.
* JSTests/wasm/gc-spec-harness/wasm-constants.js: Added.
(bytes):
(bytesWithHeader):
(makeSig):
(makeSig_v_x):
(makeSig_v_xx):
(makeSig_r_v):
(makeSig_r_x):
(makeSig_r_xx):
(assertTraps):
(assertWasmThrows):
* JSTests/wasm/gc-spec-tests/array.wast.js: Added.
* JSTests/wasm/gc-spec-tests/i31.wast.js: Added.
* JSTests/wasm/gc/const-exprs-flag-off.js: Added.
(async testInvalidGCConstExprs):
* JSTests/wasm/gc/const-exprs.js: Added.
(async testGCConstExprs):
(async testInvalidConstExprs):
* Source/JavaScriptCore/wasm/WasmConstExprGenerator.cpp:
(JSC::Wasm::ConstExprGenerator::ConstExprValue::ConstExprValue):
(JSC::Wasm::ConstExprGenerator::ConstExprValue::getValue):
(JSC::Wasm::ConstExprGenerator::ConstExprValue::type):
(JSC::Wasm::ConstExprGenerator::ConstExprValue::operator+):
(JSC::Wasm::ConstExprGenerator::ConstExprValue::operator-):
(JSC::Wasm::ConstExprGenerator::ConstExprValue::operator*):
(JSC::Wasm::ConstExprGenerator::result const):
(JSC::Wasm::ConstExprGenerator::addConstant):
(JSC::Wasm::ConstExprGenerator::getGlobal):
(JSC::Wasm::ConstExprGenerator::addRefI31):
(JSC::Wasm::ConstExprGenerator::createNewArray):
(JSC::Wasm::ConstExprGenerator::addArrayNew):
(JSC::Wasm::ConstExprGenerator::addArrayNewDefault):
(JSC::Wasm::ConstExprGenerator::addArrayNewFixed):
(JSC::Wasm::ConstExprGenerator::createNewStruct):
(JSC::Wasm::ConstExprGenerator::addStructNewDefault):
(JSC::Wasm::ConstExprGenerator::addStructNew):
(JSC::Wasm::ConstExprGenerator::addAnyConvertExtern):
(JSC::Wasm::ConstExprGenerator::addExternConvertAny):
(JSC::Wasm::ConstExprGenerator::addRefFunc):
(JSC::Wasm::evaluateExtendedConstExpr):
* Source/JavaScriptCore/wasm/WasmConstExprGenerator.h:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseExpression):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseInitExpr):
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: https://commits.webkit.org/269484@main
